### DISCUSSION TITLE
FDRS platform with light sensor (TSL2561)

### DISCUSSION THREAD

**Discussion by javrolijk** [timmbogner/Farm-Data-Relay-System#103] [Ideas]
Hi folks,

I know this was submitted in a previous post, and Tim might already be aware of the light sensor (TSL2561) request and use in the FDRS platform. Probably to get ample support, it might be worth starting a new discussion on this.

I was wondering if anyone could provide any suggestions and aid with coding. 
The plan is to place several LoRas-based sensor nodes consisting of these light sensors (TSL2561) to monitor the sky quality in our local national park and measure the developing light pollution, which you may know impacts flora and fauna. Kind of as a so-called 'Sky Quality Meter (SQM)". That way, factual data can be recorded and presented to the local authorities to take action. As you can imagine, the sensors need to be placed in remote areas in different places where Wifi/internet is unavailable, so LoRa seems to provide that solution. 
I'm not a coder, but I can understand and tailor most codes to integrate a solution for this purpose with some help.

The following table provides how the luminance (lux) correlates to the SQM value.

![image](https://user-images.githubusercontent.com/70583691/187556744-b909de66-b696-4500-ae70-1e0b5c210be4.png)

Any help, suggestions, or ideas is much appreciated.

Jairo

**Comment by timmbogner** [comment]
Sorry about that, it accidentally fell off my to-do list!

I think this should do it. It uses the Adafruit TSL2561 library, which is the first result in the libraries manager.[ Here is a link ](https://github.com/timmbogner/Farm-Data-Relay-System/tree/dev/examples/Sensor_Examples/TSL2561)to the sketch, currently in the dev branch of the repo. I have not tested it with hardware.

```
#include <Wire.h>
#include <Adafruit_Sensor.h>
#include <Adafruit_TSL2561_U.h>

#include "fdrs_sensor_config.h"
#include <fdrs_sensor.h>
   
Adafruit_TSL2561_Unified tsl = Adafruit_TSL2561_Unified(TSL2561_ADDR_FLOAT, 12345);


/**************************************************************************/
/*
    Configures the gain and integration time for the TSL2561
*/
/**************************************************************************/
void configureSensor(void)
{
  /* You can also manually set the gain or enable auto-gain support */
  // tsl.setGain(TSL2561_GAIN_1X);      /* No gain ... use in bright light to avoid sensor saturation */
  // tsl.setGain(TSL2561_GAIN_16X);     /* 16x gain ... use in low light to boost sensitivity */
  tsl.enableAutoRange(true);            /* Auto-gain ... switches automatically between 1x and 16x */
  
  /* Changing the integration time gives you better sensor resolution (402ms = 16-bit data) */
  tsl.setIntegrationTime(TSL2561_INTEGRATIONTIME_13MS);      /* fast but low resolution */
  // tsl.setIntegrationTime(TSL2561_INTEGRATIONTIME_101MS);  /* medium resolution and speed   */
  // tsl.setIntegrationTime(TSL2561_INTEGRATIONTIME_402MS);  /* 16-bit data but slowest conversions */

}

/**************************************************************************/
/*
    Arduino setup function (automatically called at startup)
*/
/**************************************************************************/
void setup(void) 
{
   beginFDRS();
  /* Initialise the sensor */
  //use tsl.begin() to default to Wire, 
  //tsl.begin(&Wire2) directs api to use Wire2, etc.
  if(!tsl.begin())
  {
    DBG("Ooops, no TSL2561 detected ... Check your wiring or I2C ADDR!");
    while(1);
  }
  configureSensor();
}

void loop(void) 
{  
  sensors_event_t event;
  tsl.getEvent(&event);
  if (event.light)
  {
    DBG(String(event.light) + " lux");
    loadFDRS(float(event.light), LIGHT_T);
    sendFDRS();
    sleepFDRS(10);  //Sleep time in seconds
  }
  else
  {
    /* If event.light = 0 lux the sensor is probably saturated
       and no reliable data could be generated! */
    Serial.println("Sensor overload");
  }
  delay(250);
}
```

**Reply by javrolijk** [reply]
Tim, no worries.
Many thanks for the prompt response, appreciate it.
I'll start testing the script. 

Jairo

**Reply by nglessner** [reply]
Jairo, you might be interested in my project which is very close to what you're wanting. The only differences I can see is that I'm using ESPNow and the TSL2591 (very similar sensor).  I've spent some time optimizing battery life which is the part you might be most interested in. At this point I have a fairly large (4400mah) battery, but it should last multiple years (using ESPNow).  I'm not sure if LoRa would end up better/worse power consumption wise.  You would have to add an external LoRa Module to what I've done.

This is the [repo](https://github.com/nglessner/Auxin).  I'm using the [DFRobot ESP32-E](https://wiki.dfrobot.com/FireBeetle_Board_ESP32_E_SKU_DFR0654) which does pretty well at low power.

You may or may not want the internal humidity/temperature sensor.  I added that to help make sure the enclosure is watertight.

**Reply by javrolijk** [reply]
> Jairo, you might be interested in my project which is very close to what you're wanting. The only differences I can see is that I'm using ESPNow and the TSL2591 (very similar sensor). I've spent some time optimizing battery life which is the part you might be most interested in. At this point I have a fairly large (4400mah) battery, but it should last multiple years (using ESPNow). I'm not sure if LoRa would end up better/worse power consumption wise. You would have to add an external LoRa Module to what I've done.
> 
> This is the [repo](https://github.com/nglessner/Auxin). I'm using the [DFRobot ESP32-E](https://wiki.dfrobot.com/FireBeetle_Board_ESP32_E_SKU_DFR0654) which does pretty well at low power.
> 
> You may or may not want the internal humidity/temperature sensor. I added that to help make sure the enclosure is watertight.

Thanks for the info.
I'll take a look at your project for sure.
